app/src/nonFree/java/com/zeapo/pwdstore/autofill/oreo/ui/AutofillSmsActivity.kt

```diff
@@ -0,0 +1,148 @@
+import com.zeapo.pwdstore.utils.viewBinding
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancelChildren
+import kotlinx.coroutines.launch
+
+@RequiresApi(Build.VERSION_CODES.O)
+class AutofillSmsActivity : AppCompatActivity(), CoroutineScope {
+
+    companion object {
+
+        private var fillOtpFromSmsRequestCode = 1
+
+        fun shouldOfferFillFromSms(context: Context): Boolean {
+            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.P)
+                return false
+            val googleApiAvailabilityInstance = GoogleApiAvailability.getInstance()
+            val googleApiStatus = googleApiAvailabilityInstance.isGooglePlayServicesAvailable(context)
+            if (googleApiStatus != ConnectionResult.SUCCESS) {
+                w { "Google Play Services unavailable or not updated: ${googleApiAvailabilityInstance.getErrorString(googleApiStatus)}" }
+                return false
+            }
+            // https://developer.android.com/guide/topics/text/autofill-services#sms-autofill
+            if (googleApiAvailabilityInstance.getApkVersion(context) < 190056000) {
+                w { "Google Play Service 19.0.56 or higher required for SMS OTP Autofill" }
+                return false
+            }
+            return true
+        }
+
+        fun makeFillOtpFromSmsIntentSender(context: Context): IntentSender {
+            val intent = Intent(context, AutofillSmsActivity::class.java)
+            return PendingIntent.getActivity(
+                context,
+                fillOtpFromSmsRequestCode++,
+                intent,
+                PendingIntent.FLAG_CANCEL_CURRENT
+            ).intentSender
+        }
+    }
+
+    private val binding by viewBinding(ActivityOreoAutofillSmsBinding::inflate)
+
+    private lateinit var clientState: Bundle
+
+    override val coroutineContext
+        get() = Dispatchers.IO + SupervisorJob()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(binding.root)
+        setResult(RESULT_CANCELED)
+        binding.cancelButton.setOnClickListener {
+            finish()
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+        clientState = intent?.getBundleExtra(AutofillManager.EXTRA_CLIENT_STATE) ?: run {
+            e { "AutofillSmsActivity started without EXTRA_CLIENT_STATE" }
+            finish()
+            return
+        }
+
+        registerReceiver(smsCodeRetrievedReceiver, IntentFilter(SmsCodeRetriever.SMS_CODE_RETRIEVED_ACTION), SmsRetriever.SEND_PERMISSION, null)
+        launch {
+            waitForSms()
+        }
+    }
+
+    // Retry starting the SMS code retriever after a permission request.
+    @Suppress("DEPRECATION")
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
```

Comment:
Let's just move this to `ActivityResultContracts`.

Reply:
We could, but would have to rely on internals of the GPS API. A `ResolvableApiException` exposes an `IntendSender` we could use, but the API docs explicitly want us to use `e.startResolutionForResult`, which does not give us access to the raw intent.

Reply:
Pity :(